Add PayPal Android SDK integration guide (#429)
paypal/android-sdk: Docs/IntegrationGuide/draft.md

# PayPal Android SDK Integration Guide

## Overview

> This guide shows you how to accept card payments and PayPal payments in your Android app using the PayPal Android SDK. You learn how to set up the SDK, initialize clients, process one-time payments, and vault payment methods for future use. Before you begin, review the Prerequisites section to make sure your environment is ready.

Reviewer: ```suggestion
This guide shows you how to accept card payments and PayPal payments in your Android app using the PayPal Android SDK. You learn how to set up the SDK, process one-time payments, and vault payment methods for future use. Before you begin, review the Prerequisites section to make sure your environment is ready.
```
I would remove "initialize clients" here because It isn't a standard term. "Set up the SDK" will be enough on its own.

---

## Prerequisites

Before you integrate the SDK, make sure you have the following:

- A PayPal developer account at [developer.paypal.com](https://developer.paypal.com)
- A sandbox client ID from the [PayPal Developer Dashboard](https://developer.paypal.com/api/rest/#link-getstarted)
- A server-side integration that can create PayPal orders and capture or authorize them (see [Orders v2 API](https://developer.paypal.com/docs/api/orders/v2/))
- For card vaulting or PayPal vaulting: a server that can create setup tokens and payment tokens (see [PayPal vaulting documentation](https://developer.paypal.com/docs/))
- Android Studio with a project targeting Android SDK 23 or higher
- Kotlin or Java project (the SDK supports both languages)

---

## Project setup

### Add SDK dependencies

Add the modules you need to your app-level `build.gradle` file. Use the current release version.

```groovy
dependencies {
    // Required: add at least one payment module

>     // For card payments (card-present and card-not-present)

Reviewer: ```suggestion
    // For card payments
```
I'd remove the card present, card not present distinction here. I believe these terms are technically meant for physical point of sale terminals.

    implementation 'com.paypal.android:card-payments:2.3.0'

    // For PayPal web checkout and PayPal vaulting
    implementation 'com.paypal.android:paypal-web-payments:2.3.0'

    // Optional: collect device data to help reduce fraud risk
    implementation 'com.paypal.android:fraud-protection:2.3.0'

    // Optional: PayPal payment button UI components
    implementation 'com.paypal.android:payment-buttons:2.3.0'

    // Optional: Venmo payments

>     implementation 'com.paypal.android:venmo:2.3.0'

Reviewer: ```suggestion
```
This can be removed until we have official support for Venmo.

}
```

All modules are published to Maven Central under the `com.paypal.android` group.

**Snapshot builds**

To test upcoming features before they ship, use snapshot builds. First, add the snapshot repository to your top-level `build.gradle`:

```groovy
repositories {
    maven {
        url 'https://central.sonatype.com/repository/maven-snapshots/'
    }
}
```

> Then reference the snapshot version:

Reviewer: ```suggestion
Then reference the latest snapshot version:
```

```groovy
implementation 'com.paypal.android:card-payments:2.3.0-SNAPSHOT'
```

### Configure build settings

> Do not hardcode your client ID in source code. Store it in a `paypal.properties` file that you exclude from version control, then inject it at build time.

Reviewer 2: This is not necessary, it's upto merchant apps, on how to handle injecting clientId,
it can be build config, a properties file, or any other way 
Same below also, all the properties related info is not needed

Reviewer: Yeah `CLIENT_ID` doesn't require the same protections as a `CLIENT_SECRET` does. They can place `CLIENT_ID` directly into their Android codebase without any issues.

Create `paypal.properties` in your project root (add this file to `.gitignore`):

```properties
CLIENT_ID=your-sandbox-client-id-here
```

In your app-level `build.gradle`, load the properties file and inject the value as a `BuildConfig` field:

```groovy
def paypalProperties = loadPropertiesFromFile("paypal.properties")

android {
    defaultConfig {
        buildConfigField("String", "CLIENT_ID", paypalProperties["CLIENT_ID"] ?: "\"\"")
    }
}

def loadPropertiesFromFile(filePath) {
    def result = new Properties()
    try {
        result.load(new FileInputStream(rootProject.file(filePath)))
    } catch (e) { /* file not found; ignore */ }
    return result
}
```

> Access the client ID at runtime using `BuildConfig.CLIENT_ID`.

Reviewer 2: upto this line, all info is regarding how to inject Ciient Id, which merchants apps can chose however they want, it can be flavors, properties file or settings

Author: @kgangineni so are you recommending that lines 77-106 are just removed? Or more?

Reviewer 2: Ya I will say remove it, all Sdk needs is `CoreConfig` how to create it is covered in other sections

### Configure deep links

> PayPal web checkout and PayPal vault flows use a browser-switch — the SDK opens a Chrome Custom Tab or the PayPal app, and PayPal redirects the customer back to your app when they finish. Be sure to configure your app to receive this redirect.

Reviewer 2: browser-switch is the name of a dependency internal to our repo; this doesn't mean anything to outside our team

Author: Interesting feedback - I kind of assumed that Browser-Switch made sense outside of the repo context. How would you recommend we describe this?

Reviewer 2: remove this part "PayPal web checkout and PayPal vault flows use a browser-switch", rest all is good,

Reviewer: Yeah we've migrated away from the Braintree `browser-switch` library. We use Chrome Custom Tabs directly in our SDK.

The SDK supports two return URL strategies. Choose one or configure both as a fallback:

**Option 1: Custom URL Scheme**

Register a custom URL scheme in your `AndroidManifest.xml`. Your activity needs to use `android:launchMode="singleTop"` so that the OS delivers the return intent to the existing activity instance rather than creating a new one.

```xml
<activity
    android:name=".MainActivity"
    android:exported="true"
    android:launchMode="singleTop">

    <!-- Standard launcher intent-filter -->
    <intent-filter>
        <action android:name="android.intent.action.MAIN" />
        <category android:name="android.intent.category.LAUNCHER" />
    </intent-filter>

>     <!-- Custom URL scheme for browser-switch return -->

Reviewer 2: better for this to be
 <!-- Custom URL scheme for return from Chrome custom tab -->

Author: Can you explain a bit more? I'm not sure I understand what your feedback is here

Reviewer 2: better for this to be 
`<!-- Custom URL scheme for returning to app -->`

Reviewer: ```suggestion
    <!-- Custom URL scheme for returning to app -->
```

    <intent-filter>
        <action android:name="android.intent.action.VIEW"/>
        <data android:scheme="com.example.myapp"/>
        <category android:name="android.intent.category.DEFAULT"/>
        <category android:name="android.intent.category.BROWSABLE"/>
    </intent-filter>
</activity>
```

Use this scheme as your `fallbackUrlScheme` in SDK requests: `"com.example.myapp"`.

**Option 2: Android App Links (HTTPS)**

App Links use HTTPS URLs with domain verification (`android:autoVerify="true"`). This requires you to host a `/.well-known/assetlinks.json` file on your domain.

```xml
<intent-filter android:autoVerify="true">
    <action android:name="android.intent.action.VIEW" />
    <category android:name="android.intent.category.DEFAULT" />
    <category android:name="android.intent.category.BROWSABLE" />
    <data
        android:host="your-app-domain.example.com"
        android:pathPrefix="/"
        android:scheme="https" />
</intent-filter>
```

Use your HTTPS URL as the `appLinkUrl` in SDK requests: `"https://your-app-domain.example.com"`.

> > **Architecture warning:** Only the activity that declares these intent-filters can receive browser-switch returns. Based on the Demo app pattern in `Demo/src/main/AndroidManifest.xml`, a single activity (`MainActivity`) owns both intent-filters. If your app has multiple activities, route all SDK flows through the activity that declares the deep link intent-filters.

Reviewer: ```suggestion
> **Architecture warning:** Only the activity that declares these intent-filters can receive deep links back into the host application. If your app has multiple activities, route all SDK flows through the activity that declares the deep link intent-filters.
```

---

## App architecture

> ### Recommended architecture

Reviewer 2: Not sure if this section is necessary, or can rename the section to Demo app architecture rather than recommended architecture, because in reality, apps are much bigger than our demo app

Reviewer: We also have two Demo apps so we might need to clarify to merchants that the merchant facing demo app is the one they should be looking at for reference.

The Demo app uses Jetpack Compose Navigation with a single entry-point activity. All SDK flows live within `MainActivity`, which hosts a `NavHost`. Individual payment screens are composable destinations. Deep link returns arrive at `MainActivity` and are dispatched to the active ViewModel.

> This single-entry-point pattern is recommended because it avoids routing ambiguity: the OS delivers deep link intents to `MainActivity`, which is the only activity that declares the browser-switch intent-filters.

Reviewer: ```suggestion
This single-entry-point pattern is recommended because it avoids routing ambiguity: the OS delivers deep link intents to `MainActivity`, which is the only activity that declares the intent-filters for deep linking.
```

            DemoApp()
        }
    }
}
```

### Process death recovery

Android may stop your app process while the customer is in the external browser completing authentication. When the customer returns, Android launches a fresh process and delivers the deep link intent. If your SDK client objects were destroyed during the process stop, they have no auth state and cannot process the return.

The SDK provides `instanceState` and `restore()` on both `CardClient` and `PayPalWebCheckoutClient` to handle this case.

**Save state before browser-switch:**

```kotlin
// In your ViewModel or activity
val savedCardClientState = cardClient.instanceState
// Persist this string (for example, in SavedStateHandle or SharedPreferences)
savedStateHandle["card_client_state"] = savedCardClientState
```

**Restore state after process death:**

```kotlin
// When recreating your client (for example, in ViewModel init)
val cardClient = CardClient(context, coreConfig)
savedStateHandle.get<String>("card_client_state")?.let { savedState ->
    cardClient.restore(savedState)
}
```

The same `instanceState` / `restore()` pattern applies to `PayPalWebCheckoutClient`.

---

> ## Initialize the SDK

Reviewer: I feel like this section could me moved up, at least before Process Death is mentioned.

> Create a `CoreConfig` object with your client ID and environment. Then create the module clients you need. The Demo app creates clients inside ViewModel constructors so they are scoped to the screen lifecycle.

Reviewer: ```suggestion
Create a `CoreConfig` object with your client ID and environment. Then create a client for each payment method you need. The Demo app creates clients inside ViewModel constructors so they are scoped to the screen lifecycle.
```

```kotlin
import com.paypal.android.corepayments.CoreConfig
import com.paypal.android.corepayments.Environment
import com.paypal.android.cardpayments.CardClient
import com.paypal.android.paypalwebpayments.PayPalWebCheckoutClient

> import com.paypal.android.fraudprotection.PayPalDataCollector

Reviewer: We should consider omitting the import statements. They will be automatically added by the Android Studio IDE. 
```suggestion
```

### Server call

Your server creates a PayPal order and returns the order ID to your app. Call your server endpoint before starting the card flow. See the [Orders v2 API](https://developer.paypal.com/docs/api/orders/v2/) for the server-side implementation.

### SDK method call

```kotlin
import com.paypal.android.cardpayments.Card
import com.paypal.android.cardpayments.CardRequest
import com.paypal.android.cardpayments.CardApproveOrderResult
import com.paypal.android.cardpayments.CardPresentAuthChallengeResult
import com.paypal.android.cardpayments.threedsecure.SCA

// Build the Card object from customer input
val card = Card(
    number = "4111111111111111",          // customer's card number
    expirationMonth = "01",               // 2-digit month (MM)
    expirationYear = "2026",              // 4-digit year (YYYY)
    securityCode = "123",                 // CVV/CVC
    cardholderName = "Jane Doe",          // optional
    billingAddress = null                 // optional
)

// Build the request
// returnUrl must match a URL scheme registered in your AndroidManifest.xml

> val returnUrl = "com.example.myapp://"   // or your HTTPS app link URL

Reviewer: ```suggestion
val returnUrl = "com.example.myapp://"    // or your HTTPS app link URL
```

val cardRequest = CardRequest(
    orderId = orderId,                    // order ID from your server
    card = card,
    returnUrl = returnUrl,
    sca = SCA.SCA_WHEN_REQUIRED           // default: only trigger 3D Secure when required
)

// Call approveOrder — result arrives asynchronously on the main thread
cardClient.approveOrder(cardRequest) { result ->
    when (result) {
        is CardApproveOrderResult.Success -> {
            // No 3D Secure challenge was needed; tell your server to capture/authorize
            val orderStatus = result.status   // nullable String — check before use

>             val attempted3DS = result.didAttemptThreeDSecureAuthentication

Reviewer: ```suggestion
```
We may not need these property references in the example since they aren't used. The `orderId` property is the main thing that matters here for capture/authorize, and it is non-null.

```kotlin
// Optional: collect device data to help reduce fraud risk before capturing
val dataCollectorRequest = PayPalDataCollectorRequest(
    hasUserLocationConsent = false  // set to true only if user consented to location data
)
val clientMetadataId = payPalDataCollector.collectDeviceData(context, dataCollectorRequest)

// Tell your server to capture/authorize
// Pass clientMetadataId in the PayPal-Client-Metadata-Id header on your server request
yourServer.captureOrder(orderId, clientMetadataId)
```

### Deep link return

When 3D Secure authentication is required, the SDK opens a Chrome Custom Tab. After the customer completes authentication, PayPal redirects them back to your app using the `returnUrl` you provided.

Register two effects in your composable to handle the return. Both are needed because Android can deliver the intent in two ways depending on whether the activity was in the foreground.

```kotlin
import com.paypal.android.cardpayments.CardFinishApproveOrderResult

// In your composable screen

> OnLifecycleOwnerResumeEffect {

Reviewer: Both `OnLifecycleOwnerResumeEffect` and `OnNewIntentEffect` are only defined in our demo app. They are not standard Android library calls. This would result in a compilation error for the merchant most likely. We could provide those helper methods in this guide. Both are less than 10 lines of code.

    // Called when the activity returns to the foreground (covers the common case)
    val intent = context.getActivityOrNull()?.intent

>     intent?.let { viewModel.completeAuthChallenge(it) }

Reviewer 2: this is not wrong, but `completeAuthChallenge` can go anywhere doesn't necessarily be in `ViewModel`, I think claude here again just following what's in demo app, instead I will put it more like 
After returning to app, use SDK API to complete transation using 
```kotlin
 cardClient.finishApproveOrder(intent)?.let { result ->
        // finishApproveOrder returns null if the intent is not for this auth session
        when (result) {
            is CardFinishApproveOrderResult.Success -> {
                // Authentication succeeded; capture/authorize the order on your server
                val orderStatus = result.status  // nullable — check before use
// make server call to capture order in your server
            }

            is CardFinishApproveOrderResult.Failure ->
// Handle UI for failure scenarios

            CardFinishApproveOrderResult.Canceled ->
                // Handle UI for Cancelled scenarios

            CardFinishApproveOrderResult.NoResult -> {
                // This intent was not for this auth session
                // Re-enable the pay button so the customer can retry
                resetToIdleState()
            }
        }
    }
```

}

OnNewIntentEffect { newIntent ->
    // Called when a new intent arrives at the singleTop activity
    viewModel.completeAuthChallenge(newIntent)
}

// In your ViewModel
fun completeAuthChallenge(intent: Intent) {
    cardClient.finishApproveOrder(intent)?.let { result ->
        // finishApproveOrder returns null if the intent is not for this auth session
        when (result) {
            is CardFinishApproveOrderResult.Success -> {
                // Authentication succeeded; capture/authorize the order on your server

>                 val orderStatus = result.status  // nullable — check before use

Reviewer: ```suggestion
```
We can remove this property reference as well. Technically the enum `.Success` type itself is the status.

    when (startResult) {
        is PayPalPresentAuthChallengeResult.Success -> {
            // Chrome Custom Tab or PayPal app opened
            // Do nothing here — wait for the deep link return
        }

        is PayPalPresentAuthChallengeResult.Failure ->
            showError(startResult.error.errorDescription)
    }
}
```

`PayPalWebCheckoutFundingSource` options:

| Value | Description |
|---|---|
| `PAYPAL` | Standard PayPal checkout (default) |
| `PAYPAL_CREDIT` | PayPal Credit — revolving line of credit for eligible customers |
| `PAY_LATER` | Pay Later — short-term, interest-free payments for eligible customers |

### Handle the result

After the flow completes, capture or authorize the order on your server.

```kotlin
// After finishStart returns Success
fun capturePayPalOrder(orderId: String?) {
    if (orderId == null) {
        showError("Order ID is missing")
        return
    }
    yourServer.captureOrder(orderId)
}
```

### Deep link return

Register both effects in your composable. Call `finishStart()` in your ViewModel.

```kotlin
import com.paypal.android.paypalwebpayments.PayPalWebCheckoutFinishStartResult

// In your composable

> OnLifecycleOwnerResumeEffect {

Reviewer 2: this is how our demo app does
`OnLifecycleOwnerResume`, `onNewIntentEffect`, but merchant apps can chose however they want, it's better to say it override `OnResume`, `OnNewIntent` functions in activity to handle return to app, same comment above also

        }
    }
}
```

### What's next

After creating the payment token on your server, store the token ID for future charges.

---

## Testing

Use the sandbox environment during development. Set `environment = Environment.SANDBOX` in your `CoreConfig`.

The Demo app connects to a hosted sample server for testing purposes. Your integration connects to your own server, which calls the PayPal sandbox APIs.

**Test sandbox accounts**

Create sandbox test accounts in the [PayPal Developer Dashboard](https://developer.paypal.com) under your sandbox environment. Use these accounts when logging in during PayPal web checkout tests.

**Test scenarios to verify**

- Card payment succeeds without 3D Secure
- Card payment triggers 3D Secure challenge — customer completes authentication
- Card payment triggers 3D Secure challenge — customer cancels
- PayPal checkout completes
- PayPal checkout — customer cancels in Chrome Custom Tab
- PayPal checkout — app switch to PayPal app (if testing `appSwitchWhenEligible = true`)
- Process death recovery — stop the app while authentication is in progress, return to it, and verify the flow completes correctly

**Test process death recovery**

> 1. Start a checkout flow that requires browser-switch (PayPal web checkout or card 3D Secure).

Reviewer: ```suggestion
1. Start a checkout flow that requires authentication via Chrome Custom Tab (PayPal web checkout or card 3D Secure).
```

2. After the Chrome Custom Tab opens, use Android Studio's "Stop app" button to stop the process.
3. Return to the app by selecting the notification or using the back button from the browser.
4. Verify that your `restore()` call rebuilds the SDK client state and that `finishStart()` or `finishApproveOrder()` returns the correct result.

---

## Go live checklist

Before going live, complete the following:

- [ ] Replace `Environment.SANDBOX` with `Environment.LIVE` in your `CoreConfig`

> - [ ] Replace your sandbox client ID with your live client ID in `paypal.properties`

Reviewer: ```suggestion
- [ ] Replace your sandbox client ID with your live client ID
```
